lib/kino_slack/message_interpolator.ex

```diff
@@ -0,0 +1,53 @@
+defmodule KinoSlack.MessageInterpolator do
+  @moduledoc false
+
+  def interpolate(message) do
+    args = build_interpolation_args(message, [])
+    {:<<>>, [], args}
+  end
+
+  defp build_interpolation_args("", args) do
+    args
+  end
+
+  defp build_interpolation_args("{{" <> rest, args) do
+    with [inner, rest] <- String.split(rest, "}}", parts: 2),
+         {:ok, expression} <- Code.string_to_quoted(inner) do
+      args = append_interpolation(args, expression)
+      build_interpolation_args(rest, args)
+    else
+      _ ->
+        args = append_string(args, "{{")
+        build_interpolation_args(rest, args)
+    end
+  end
+
+  defp build_interpolation_args(<<char::utf8, rest::binary>>, args) do
+    args = append_string(args, <<char::utf8>>)
+    build_interpolation_args(rest, args)
+  end
+
+  defp append_interpolation(args, expression) do
+    interpolation_node = {
+      :"::",
+      [],
+      [
+        {{:., [], [Kernel, :to_string]}, [], [expression]},
+        {:binary, [], Elixir}
+      ]
+    }
+
+    args ++ [interpolation_node]
```

Comment:
You want to avoid appending to the end of the list, because it is expensive. The idea is that you always prepend and at the end you call `Enum.reverse`. This will also make operations such as `append_string` faster and easier, because you only need to access the head of the list. :)

Reply:
Done. ✅